frontend/integration-tests/views/login.view.ts

```diff
@@ -1,23 +1,19 @@
 import { $, browser, ExpectedConditions as until, by, element } from 'protractor';
-import { appHost } from '../protractor.conf';
 
 export const nameInput = $('#inputUsername');
 export const passwordInput = $('#inputPassword');
 export const submitButton = $('button[type=submit]');
 export const logOutLink = element(by.linkText('Log out'));
 export const userDropdown = $('[data-test=user-dropdown] .pf-c-dropdown__toggle');
+export const loginPF = $('.login-pf');
 
 export const selectProvider = async(provider: string) => {
   const idpLink = element(by.cssContainingText('.idp', provider));
-  while (!(await idpLink.isPresent())) {
```

Comment:
I don't think we can remove this. At the beginning of our integration tests, we create an htpasswd idp, which takes some time to propagate and be seen on the login page. The login page is static, so if the idp creation has not propagated through the system when the integration tests get to this point, then we won't get redirected to the idp selection page, and the `idpLink` element won't exist. We have to keep trying to visit the console base url until the idp has been created, and we are redirected to the idp selection page. If our test case tries to log in with a provider that doesn't exist, or tries to explicitly select a provider when it is the only one present (i.e. we won't ever see the idp selection page), the test will time out at this step and fail, which is what we would want.

If there is only one idp (the kubeadmin provider) on the cluster, then we can omit the `provider` arg from the call to the `login` helper function.

Reply:
@TheRealJon Could you show me which part or test file is responsibling for creating the htpasswd idp, I thought it's been created via the cluster deployment. If it need to keep checking the `idp` presence, there is no way to do login in `login` function for only one idp (the kubeadmin provider), as it has no `idp`.  The only way is call the `login` without `provider` as you said, but there is no condition can be used to set for when to omit `provider`, unless use a environment variable to control it.

The ideal situation to me is:
1. if there is no `idp`, skip `selectProvider` in `login`.
2. if there are `idp`, which mean they're ready, these lines no needs any more.

Reply:
Hi @gouyang, sorry this slipped through the cracks. The idp is created in our `test-prow-e2e.sh` script. This adds a complication to the 2 scenarios you provide as your ideal situation. There is a third scenario where an idp has been created, but has not propagated through the system yet, and therefore won't show up in the login workflow yet. This will almost always be the case in our e2e tests since the idp is created just before the tests run.